// sveld/lib/writer/writer-markdown.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
exports.__esModule = true;
var path = require("path");
var WriterMarkdown_1 = require("./WriterMarkdown");
var writer_ts_definitions_1 = require("./writer-ts-definitions");
var PROP_TABLE_HEADER = "| Prop name | Kind | Reactive | Type | Default value | Description |\n| :- | :- | :- | :- |\n";
var SLOT_TABLE_HEADER = "| Slot name | Default | Props | Fallback |\n| :- | :- | :- | :- |\n";
var EVENT_TABLE_HEADER = "| Event name | Type | Detail |\n| :- | :- | :- |\n";
var MD_TYPE_UNDEFINED = "--";
function formatPropType(type) {
    if (type === undefined)
        return MD_TYPE_UNDEFINED;
    return "<code>" + type.replace(/\|/g, "&#124;") + "</code>";
}
function escapeHtml(text) {
    return text.replace(/\</g, "&lt;").replace(/\>/g, "&gt;");
}
function formatPropValue(value) {
    if (value === undefined)
        return MD_TYPE_UNDEFINED;
    return "<code>" + value.replace(/`/g, "\\`").replace(/\|/g, "&#124;") + "</code>";
}
function formatPropDescription(description) {
    if (description === undefined || description.trim().length === 0)
        return MD_TYPE_UNDEFINED;
    return escapeHtml(description).replace(/\n/g, "<br />");
}
function formatSlotProps(props) {
    if (props === undefined || props === "{}")
        return MD_TYPE_UNDEFINED;
    return formatPropType(writer_ts_definitions_1.formatTsProps(props).replace(/\n/g, " "));
}
function formatSlotFallback(fallback) {
    if (fallback === undefined)
        return MD_TYPE_UNDEFINED;
    return formatPropType(escapeHtml(fallback).replace(/\n/g, "<br />"));
}
function formatEventDetail(detail) {
    if (detail === undefined)
        return MD_TYPE_UNDEFINED;
    return formatPropType(detail.replace(/\n/g, " "));
}
function writeMarkdown(components, options) {
    return __awaiter(this, void 0, void 0, function () {
        var output_path, document, keys;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    output_path = path.join(process.cwd(), options.outFile);
                    document = new WriterMarkdown_1["default"]({
                        onAppend: function (type, document) {
                            var _a;
                            (_a = options.onAppend) === null || _a === void 0 ? void 0 : _a.call(null, type, document, components);
                        }
                    });
                    document.append("h1", "Component Index");
                    document.append("h2", "Components").tableOfContents();
                    document.append("divider");
                    keys = Array.from(components.keys()).sort();
                    keys.forEach(function (key) {
                        var component = components.get(key);
                        document.append("h2", "`" + component.moduleName + "`");
                        if (component.typedefs.length > 0) {
                            document.append("h3", "Types").append("raw", "```ts\n" + writer_ts_definitions_1.getTypeDefs({
                                typedefs: component.typedefs
                            }) + "\n```\n\n");
                        }
                        document.append("h3", "Props");
                        if (component.props.length > 0) {
                            document.append("raw", PROP_TABLE_HEADER);
                            __spreadArrays(component.props).sort(function (a) {
                                if (a.reactive)
                                    return -1;
                                if (a.constant)
                                    return 1;
                                return 0;
                            })
                                .forEach(function (prop) {
                                document.append("raw", "| " + prop.name + " | " + ("<code>" + prop.kind + "</code>") + " | " + (prop.reactive ? "Yes" : "No") + " | " + formatPropType(prop.type) + " | " + formatPropValue(prop.value) + " | " + formatPropDescription(prop.description) + " |\n");
                            });
                        }
                        else {
                            document.append("p", "None.");
                        }
                        document.append("h3", "Slots");
                        if (component.slots.length > 0) {
                            document.append("raw", SLOT_TABLE_HEADER);
                            component.slots.forEach(function (slot) {
                                document.append("raw", "| " + (slot["default"] ? MD_TYPE_UNDEFINED : slot.name) + " | " + (slot["default"] ? "Yes" : "No") + " | " + formatSlotProps(slot.slot_props) + " | " + formatSlotFallback(slot.fallback) + " |\n");
                            });
                        }
                        else {
                            document.append("p", "None.");
                        }
                        document.append("h3", "Events");
                        if (component.events.length > 0) {
                            document.append("raw", EVENT_TABLE_HEADER);
                            component.events.forEach(function (event) {
                                document.append("raw", "| " + event.name + " | " + event.type + " | " + (event.type === "dispatched" ? formatEventDetail(event.detail) : MD_TYPE_UNDEFINED) + " |\n");
                            });
                        }
                        else {
                            document.append("p", "None.");
                        }
                    });
                    return [4 /*yield*/, document.write(output_path, document.end())];
                case 1:
                    _a.sent();
                    process.stdout.write("created \"" + options.outFile + "\".\n");
                    return [2 /*return*/];
            }
        });
    });
}
exports["default"] = writeMarkdown;
