// sveld/lib/ComponentParser.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
exports.__esModule = true;
var compiler_1 = require("svelte/compiler");
var commentParser = require("comment-parser");
var element_tag_map_1 = require("./element-tag-map");
var DEFAULT_SLOT_NAME = "__default__";
var ComponentParser = /** @class */ (function () {
    function ComponentParser(options) {
        this.reactive_vars = new Set();
        this.props = new Map();
        this.slots = new Map();
        this.events = new Map();
        this.typedefs = new Map();
        this.bindings = new Map();
        this.options = options;
    }
    ComponentParser.mapToArray = function (map) {
        return Array.from(map, function (_a) {
            var key = _a[0], value = _a[1];
            return value;
        });
    };
    ComponentParser.assignValue = function (value) {
        return value === undefined || value === "" ? undefined : value;
    };
    ComponentParser.formatComment = function (comment) {
        var formatted_comment = comment;
        if (!formatted_comment.startsWith("/*")) {
            formatted_comment = "/*" + formatted_comment;
        }
        if (!formatted_comment.endsWith("*/")) {
            formatted_comment += "*/";
        }
        return formatted_comment;
    };
    ComponentParser.prototype.sourceAtPos = function (start, end) {
        var _a;
        return (_a = this.source) === null || _a === void 0 ? void 0 : _a.slice(start, end);
    };
    ComponentParser.prototype.collectReactiveVars = function () {
        var _this = this;
        var _a;
        (_a = this.compiled) === null || _a === void 0 ? void 0 : _a.vars.filter(function (_a) {
            var reassigned = _a.reassigned, writable = _a.writable;
            return reassigned && writable;
        }).forEach(function (_a) {
            var name = _a.name;
            return _this.reactive_vars.add(name);
        });
    };
    ComponentParser.prototype.addProp = function (prop_name, data) {
        if (ComponentParser.assignValue(prop_name) === undefined)
            return;
        if (this.props.has(prop_name)) {
            var existing_slot = this.props.get(prop_name);
            this.props.set(prop_name, __assign(__assign({}, existing_slot), data));
        }
        else {
            this.props.set(prop_name, data);
        }
    };
    ComponentParser.prototype.aliasType = function (type) {
        if (type === "*")
            return "any";
        return type;
    };
    ComponentParser.prototype.addSlot = function (slot_name, slot_props, slot_fallback) {
        var default_slot = slot_name === undefined || slot_name === "";
        var name = default_slot ? DEFAULT_SLOT_NAME : slot_name;
        var fallback = ComponentParser.assignValue(slot_fallback);
        var props = ComponentParser.assignValue(slot_props);
        if (this.slots.has(name)) {
            var existing_slot = this.slots.get(name);
            this.slots.set(name, __assign(__assign({}, existing_slot), { fallback: fallback, slot_props: existing_slot.slot_props === undefined ? props : existing_slot.slot_props }));
        }
        else {
            this.slots.set(name, {
                name: name,
                "default": default_slot,
                fallback: fallback,
                slot_props: slot_props
            });
        }
    };
    ComponentParser.prototype.addDispatchedEvent = function (name, detail) {
        if (name === undefined)
            return;
        if (this.events.has(name)) {
            var existing_event = this.events.get(name);
            this.events.set(name, __assign(__assign({}, existing_event), { detail: existing_event.detail === undefined ? ComponentParser.assignValue(detail) : existing_event.detail }));
        }
        else {
            this.events.set(name, {
                type: "dispatched",
                name: name,
                detail: ComponentParser.assignValue(detail)
            });
        }
    };
    ComponentParser.prototype.parseCustomTypes = function () {
        var _this = this;
        commentParser(this.source).forEach(function (_a) {
            var tags = _a.tags;
            tags.forEach(function (_a) {
                var tag = _a.tag, tagType = _a.type, name = _a.name, description = _a.description;
                var type = _this.aliasType(tagType);
                switch (tag) {
                    case "extends":
                        _this["extends"] = {
                            interface: name,
                            "import": type
                        };
                        break;
                    case "restProps":
                        _this.rest_props = {
                            type: "Element",
                            name: type
                        };
                        break;
                    case "slot":
                        _this.addSlot(name, type);
                        break;
                    case "event":
                        _this.addDispatchedEvent(name, type);
                        break;
                    case "typedef":
                        _this.typedefs.set(name, {
                            type: type,
                            name: name,
                            description: ComponentParser.assignValue(description),
                            ts: /(\}|\};)$/.test(type) ? "interface " + name + " " + type : "type " + name + " = " + type
                        });
                        break;
                }
            });
        });
    };
    ComponentParser.prototype.cleanup = function () {
        this.source = undefined;
        this.compiled = undefined;
        this.rest_props = undefined;
        this["extends"] = undefined;
        this.reactive_vars.clear();
        this.props.clear();
        this.slots.clear();
        this.events.clear();
        this.typedefs.clear();
        this.bindings.clear();
    };
    ComponentParser.prototype.parseSvelteComponent = function (source, diagnostics) {
        var _this = this;
        var _a;
        if ((_a = this.options) === null || _a === void 0 ? void 0 : _a.verbose) {
            process.stdout.write("[parsing] \"" + diagnostics.moduleName + "\" " + diagnostics.filePath + "\n");
        }
        this.cleanup();
        this.source = source;
        this.compiled = compiler_1.compile(source);
        this.collectReactiveVars();
        this.parseCustomTypes();
        var dispatcher_name = undefined;
        var callees = [];
        compiler_1.walk(this.compiled.ast, {
            enter: function (node, parent, prop) {
                var _a, _b, _c, _d, _e, _f, _g, _h;
                if (node.type === "CallExpression") {
                    if (node.callee.name === "createEventDispatcher") {
                        dispatcher_name = parent === null || parent === void 0 ? void 0 : parent.id.name;
                    }
                    callees.push({
                        name: node.callee.name,
                        arguments: node.arguments
                    });
                }
                if (node.type === "Spread" && (node === null || node === void 0 ? void 0 : node.expression.name) === "$$restProps") {
                    if (_this.rest_props === undefined && ((parent === null || parent === void 0 ? void 0 : parent.type) === "InlineComponent" || (parent === null || parent === void 0 ? void 0 : parent.type) === "Element")) {
                        _this.rest_props = {
                            type: parent.type,
                            name: parent.name
                        };
                    }
                }
                if (node.type === "ExportNamedDeclaration" && node.declaration != null) {
                    var _j = node.declaration.declarations
                        ? node.declaration.declarations[0]
                        : node.declaration, declaration_type = _j.type, id = _j.id, init = _j.init, body = _j.body;
                    var prop_name = id.name;
                    var value = undefined;
                    var type = undefined;
                    var kind = node.declaration.kind;
                    var description = undefined;
                    var isFunction = false;
                    var isFunctionDeclaration = false;
                    if (init != null) {
                        if (init.type === "ObjectExpression" ||
                            init.type === "BinaryExpression" ||
                            init.type === "ArrayExpression" ||
                            init.type === "ArrowFunctionExpression") {
                            value = (_a = _this.sourceAtPos(init.start, init.end)) === null || _a === void 0 ? void 0 : _a.replace(/\n/g, " ");
                            type = value;
                            isFunction = init.type === "ArrowFunctionExpression";
                            if (init.type === "BinaryExpression") {
                                if ((init === null || init === void 0 ? void 0 : init.left.type) === "Literal" && typeof (init === null || init === void 0 ? void 0 : init.left.value) === "string") {
                                    type = "string";
                                }
                            }
                        }
                        else {
                            if (init.type === "UnaryExpression") {
                                value = _this.sourceAtPos(init.start, init.end);
                                type = typeof ((_b = init.argument) === null || _b === void 0 ? void 0 : _b.value);
                            }
                            else {
                                value = init.raw;
                                type = init.value == null ? undefined : typeof init.value;
                            }
                        }
                    }
                    if (declaration_type === "FunctionDeclaration") {
                        value = "() => " + ((_c = _this.sourceAtPos(body.start, body.end)) === null || _c === void 0 ? void 0 : _c.replace(/\n/g, " "));
                        type = "() => any";
                        kind = "function";
                        isFunction = true;
                        isFunctionDeclaration = true;
                    }
                    if (node.leadingComments) {
                        var last_comment = node.leadingComments[node.leadingComments.length - 1];
                        var comment = commentParser(ComponentParser.formatComment(last_comment.value));
                        var tag = (_d = comment[0]) === null || _d === void 0 ? void 0 : _d.tags[((_e = comment[0]) === null || _e === void 0 ? void 0 : _e.tags.length) - 1];
                        if ((tag === null || tag === void 0 ? void 0 : tag.tag) === "type")
                            type = _this.aliasType(tag.type);
                        description = ComponentParser.assignValue((_f = comment[0]) === null || _f === void 0 ? void 0 : _f.description);
                    }
                    if (!description && _this.typedefs.has(type)) {
                        description = _this.typedefs.get(type).description;
                    }
                    _this.addProp(prop_name, {
                        name: prop_name,
                        kind: kind,
                        description: description,
                        type: type,
                        value: value,
                        isFunction: isFunction,
                        isFunctionDeclaration: isFunctionDeclaration,
                        constant: kind === "const",
                        reactive: _this.reactive_vars.has(prop_name)
                    });
                }
                if (node.type === "Slot") {
                    var slot_name = (_g = node.attributes.find(function (attr) { return attr.name === "name"; })) === null || _g === void 0 ? void 0 : _g.value[0].data;
                    var slot_props = node.attributes
                        .filter(function (attr) { return attr.name !== "name"; })
                        .reduce(function (slot_props, _a) {
                        var _b;
                        var name = _a.name, value = _a.value;
                        var slot_prop_value = {
                            value: undefined,
                            replace: false
                        };
                        if (value[0]) {
                            var _c = value[0], type = _c.type, expression = _c.expression, raw = _c.raw, start = _c.start, end = _c.end;
                            if (type === "Text") {
                                slot_prop_value.value = raw;
                            }
                            else if (type === "AttributeShorthand") {
                                slot_prop_value.value = expression.name;
                                slot_prop_value.replace = true;
                            }
                            if (expression) {
                                if (expression.type === "Literal") {
                                    slot_prop_value.value = expression.value;
                                }
                                else if (expression.type !== "Identifier") {
                                    if (expression.type === "ObjectExpression" || expression.type === "TemplateLiteral") {
                                        slot_prop_value.value = _this.sourceAtPos(start + 1, end - 1);
                                    }
                                    else {
                                        slot_prop_value.value = _this.sourceAtPos(start, end);
                                    }
                                }
                            }
                        }
                        return __assign(__assign({}, slot_props), (_b = {}, _b[name] = slot_prop_value, _b));
                    }, {});
                    var fallback = (_h = node.children) === null || _h === void 0 ? void 0 : _h.map(function (_a) {
                        var start = _a.start, end = _a.end;
                        return _this.sourceAtPos(start, end);
                    }).join("").trim();
                    _this.addSlot(slot_name, JSON.stringify(slot_props, null, 2), fallback);
                }
                if (node.type === "EventHandler" && node.expression == null) {
                    if (!_this.events.has(node.name) && parent !== undefined) {
                        _this.events.set(node.name, {
                            type: "forwarded",
                            name: node.name,
                            element: parent.name
                        });
                    }
                }
                if ((parent === null || parent === void 0 ? void 0 : parent.type) === "Element" && node.type === "Binding" && node.name === "this") {
                    var prop_name = node.expression.name;
                    var element_name = parent.name;
                    if (_this.bindings.has(prop_name)) {
                        var existing_bindings = _this.bindings.get(prop_name);
                        if (!existing_bindings.elements.includes(element_name)) {
                            _this.bindings.set(prop_name, __assign(__assign({}, existing_bindings), { elements: __spreadArrays(existing_bindings.elements, [element_name]) }));
                        }
                    }
                    else {
                        _this.bindings.set(prop_name, {
                            elements: [element_name]
                        });
                    }
                }
            }
        });
        if (dispatcher_name !== undefined) {
            callees.forEach(function (callee) {
                var _a, _b;
                if (callee.name === dispatcher_name) {
                    var event_name = (_a = callee.arguments[0]) === null || _a === void 0 ? void 0 : _a.value;
                    var event_detail = (_b = callee.arguments[1]) === null || _b === void 0 ? void 0 : _b.value;
                    _this.addDispatchedEvent(event_name, event_detail);
                }
            });
        }
        return {
            props: ComponentParser.mapToArray(this.props).map(function (prop) {
                if (_this.bindings.has(prop.name)) {
                    return __assign(__assign({}, prop), { type: "null | " +
                            _this.bindings
                                .get(prop.name)
                                .elements.sort()
                                .map(function (element) { return element_tag_map_1.getElementByTag(element); })
                                .join(" | ") });
                }
                return prop;
            }),
            slots: ComponentParser.mapToArray(this.slots)
                .map(function (slot) {
                try {
                    var slot_props_1 = JSON.parse(slot.slot_props);
                    var new_props_1 = [];
                    Object.keys(slot_props_1).forEach(function (key) {
                        var _a;
                        if (slot_props_1[key].replace && slot_props_1[key].value !== undefined) {
                            slot_props_1[key].value = (_a = _this.props.get(slot_props_1[key].value)) === null || _a === void 0 ? void 0 : _a.type;
                        }
                        if (slot_props_1[key].value === undefined)
                            slot_props_1[key].value = "any";
                        new_props_1.push(key + ": " + slot_props_1[key].value);
                    });
                    var formatted_slot_props = new_props_1.length === 0 ? "{}" : "{ " + new_props_1.join(", ") + " }";
                    return __assign(__assign({}, slot), { slot_props: formatted_slot_props });
                }
                catch (e) {
                    return slot;
                }
            })
                .sort(function (a, b) {
                if (a.name < b.name)
                    return -1;
                if (a.name > b.name)
                    return 1;
                return 0;
            }),
            events: ComponentParser.mapToArray(this.events),
            typedefs: ComponentParser.mapToArray(this.typedefs),
            rest_props: this.rest_props,
            "extends": this["extends"]
        };
    };
    return ComponentParser;
}());
exports["default"] = ComponentParser;
